src/read_write/raw.rs

```diff
@@ -107,45 +109,228 @@ impl NodeReader for RawNodeReader {
+                        });
+                        batch
+                            .attributes
+                            .insert(key.to_owned(), AttributeData::U64(attr));
+                    }
+                    AttributeDataType::F32 => {
+                        let mut attr = Vec::with_capacity(NUM_POINTS_PER_BATCH);
+                        let _err: io::Result<()> = (0..NUM_POINTS_PER_BATCH).try_for_each(|_| {
+                            let data = reader.read_f32::<LittleEndian>()?;
+                            attr.push(data);
+                            Ok(())
+                        });
+                        batch
+                            .attributes
+                            .insert(key.to_owned(), AttributeData::F32(attr));
+                    }
+                    AttributeDataType::F64 => {
+                        let mut attr = Vec::with_capacity(NUM_POINTS_PER_BATCH);
+                        let _err: io::Result<()> = (0..NUM_POINTS_PER_BATCH).try_for_each(|_| {
+                            let data = reader.read_f64::<LittleEndian>()?;
+                            attr.push(data);
+                            Ok(())
+                        });
+                        batch
+                            .attributes
+                            .insert(key.to_owned(), AttributeData::F64(attr));
+                    }
+                    AttributeDataType::U8Vec3 => {
+                        let mut attr = Vec::with_capacity(NUM_POINTS_PER_BATCH);
+                        let _err: io::Result<()> = (0..NUM_POINTS_PER_BATCH).try_for_each(|_| {
+                            let x = reader.read_u8()?;
+                            let y = reader.read_u8()?;
+                            let z = reader.read_u8()?;
+                            attr.push(Vector3::new(x, y, z));
+                            Ok(())
+                        });
+                        batch
+                            .attributes
+                            .insert(key.to_owned(), AttributeData::U8Vec3(attr));
+                    }
+                    AttributeDataType::F64Vec3 => {
+                        let mut attr = Vec::with_capacity(NUM_POINTS_PER_BATCH);
+                        let _err: io::Result<()> = (0..NUM_POINTS_PER_BATCH).try_for_each(|_| {
+                            let x = reader.read_f64::<LittleEndian>()?;
+                            let y = reader.read_f64::<LittleEndian>()?;
+                            let z = reader.read_f64::<LittleEndian>()?;
+                            attr.push(Vector3::new(x, y, z));
+                            Ok(())
+                        });
+                        batch
+                            .attributes
+                            .insert(key.to_owned(), AttributeData::F64Vec3(attr));
+                    }
+                };
+            });
+
+        let num_points = batch.position.len();
+        if num_points == 0 {
+            return None;
```

Comment:
Does it hurt, if the batch is empty?

Reply:
Iterators should return None to indicate that iteration has finished, otherwise for_each etc. don't terminate.

Reply:
Okay, it's not an Iterator instance, so that doesn't hold. Hmm.